### PR TITLE
perf(vim.text): use lookup table implementation for hex encoding

### PR DESCRIPTION
**Pull request by gpanders**
Credit to @glepnir for the suggestion.

Ref: https://github.com/neovim/neovim/issues/29981#issuecomment-2295081391

**Before**:

```
Benchmarking vim.text.hexencode...
Time taken: 0.059076
Benchmarking vim.text.hexdecode...
Time taken: 0.336875
```

**After**:

```
Benchmarking vim.text.hexencode...
Time taken: 0.038409
Benchmarking vim.text.hexdecode...
Time taken: 0.084737
```

### CODE REVIEW

File: runtime/lua/vim/text.lua [hunk 1]
```diff
@@ -2,14 +2,27 @@
...
 ---
 --- @param str string String to encode
 --- @return string : Hex encoded string
 function M.hexencode(str)
   local enc = {} ---@type string[]
   for i = 1, #str do
-    enc[i] = string.format('%02X', str:byte(i, i + 1))
+    local byte = str:byte(i)
+    enc[#enc + 1] = lookup[math.floor(byte / 16)] --[[@as string]]
+    enc[#enc + 1] = lookup[byte % 16] --[[@as string]]
```
Comment (lewis6991) [inline runtime/lua/vim/text.lua:25]: Would be curious to know if calculating the correct value of `i` (and avoiding `#enc`) (like in `hexdecode()`) makes a difference here?

**Review by lewis6991 — APPROVED** [review]
If possible, the benchmarking code you used might be useful to add in `benchmark/`

File: runtime/lua/vim/text.lua [hunk 2]
```diff
@@ -2,14 +2,27 @@
 
 local M = {}
 
+local alphabet = '0123456789ABCDEF'
+local lookup = {} ---@type table<integer|string, integer|string>
```
Comment (lewis6991) [inline runtime/lua/vim/text.lua:6]: Can we split this table into two?

**Comment by Danielkonge** [comment]
Is there any reason not to encode/decode the whole byte instead of 4 bits at a time? Do we care about that small amount of extra memory used? Something like this (which might not have the best names...) is faster when I benchmarked it:
```lua
-- Text processing functions.

local alphabet = '0123456789ABCDEF'
local lookup_string = {} ---@type table<integer, string>
local lookup_integer = {} ---@type table<string, integer>
do
  for i = 1, #alphabet do
    local char1 = alphabet:sub(i, i)
    for j = 1, #alphabet do
      local char2 = alphabet:sub(j, j)
      local s = char1 .. char2
      lookup_string[16 * (i - 1) + (j - 1)] = s
      lookup_integer[s] = 16 * (i - 1) + (j - 1)
      lookup_integer[s:lower()] = 16 * (i - 1) + (j - 1)
    end
  end
end

--- Hex encode a string.
---
--- @param str string String to encode
--- @return string : Hex encoded string
local function hexencode(str)
  local enc = {} ---@type string[]
  for i = 1, #str do
    local byte = str:byte(i)
    enc[i] = lookup_string[byte]
  end
  return table.concat(enc)
end

--- Hex decode a string.
---
--- @param enc string String to decode
--- @return string? : Decoded string
--- @return string? : Error message, if any
local function hexdecode(enc)
  if #enc % 2 ~= 0 then
    return nil, 'string must have an even number of hex characters'
  end

  local str = {} ---@type string[]
  local index = 1
  for i = 1, math.floor(#enc / 2) do
    local byte = lookup_integer[enc:sub(index, index + 1)]
    if not byte then
      return nil, 'string must contain only hex characters'
    end
    str[i] = string.char(byte)
    index = index + 2
  end
  return table.concat(str), nil
end

```

**Comment by gpanders** [comment]
```
$ TEST_FILE=test/benchmark/text_spec.lua make functionaltest
-------- Global test environment setup.
-------- Running tests from test/benchmark/text_spec.lua
RUN       vim.text hexencode:
Min: 2.806411 ms, Max: 4.499787 ms, Median: 3.098527 ms, Mean: 3.124150 ms
313.00 ms OK
RUN       vim.text hexdecode:
Min: 5.642333 ms, Max: 25.032607 ms, Median: 6.953082 ms, Mean: 7.113856 ms
712.00 ms OK
-------- 2 tests from test/benchmark/text_spec.lua (1026.00 ms total)

-------- Global test environment teardown.
======== 2 tests from 1 test file ran. (1027.00 ms total)
PASSED   2 tests.
```

**Pull request merged**